# Business Goal/Outcome

## Goal
Give a realtor like Jeanine a data-backed answer to the question every seller asks ("What's my home worth, and how long will it take?") and every buyer asks ("Do I have room to negotiate?"), specific to the zip they're standing in.

## Outcomes
- Realtor can price the seller correctly in listing appts 
- Aide buyer consolutations and negotiation strategy
- Help with prospecting and knowing which zips to farm
- Data to feed market update content
- Help relay market expertise 

# Research Question
Where in Orlando have sellers lost leverage since 2022, and where do they still have it?

# Hypothesis

**H1: New construction hurt resale sellers most.**  
		In 34787 and 32712, price cut share and days on market will have risen more than the metro average since 2022, because builders' incentives pull buyers away from resale homes.  
		_Rejected if:_ both zips' increases are at or below the metro average.


**H2: Windermere held up.**  
		In 34786, the leverage score will have dropped less than the metro average since 2022, because scarce lakefront and large-lot homes and less rate-sensitive buyers protect sellers.  
		_Rejected if:_ 34786's drop is at or above the metro average. If that happens, the Horizon West portion of the zip may be dragging it down.


**H3: Affordable zips**
 34761 and 32703 will have lost less leverage than 34787 and 32712, because they have less new-construction competition. _Rejected if:_ their drop is the same as or larger than the growth zips'. 

# Bring in Data For Prep

In [235]:
import pandas as pd


In [236]:
historical = pd.read_csv('/Users/medhasubramaniyan/Desktop/GitHub_Repos/orlando-seller-leverage-analysis/data/RDC_Inventory_Core_Metrics_Zip_History.csv')
historical

,month_date_yyyymm,postal_code,zip_name,median_listing_price,median_listing_price_mm,median_listing_price_yy,active_listing_count,active_listing_count_mm,active_listing_count_yy,median_days_on_market,...,average_listing_price,average_listing_price_mm,average_listing_price_yy,total_listing_count,total_listing_count_mm,total_listing_count_yy,pending_ratio,pending_ratio_mm,pending_ratio_yy,quality_flag
0,202608,74003,"bartlesville, ok",150000.0,0.0000,0.0714,45,-0.0722,0.0714,70.0,...,208058.0,-0.0086,0.0436,71,-0.0596,0.1639,0.5778,0.0417,0.1254,0.0
1,202608,61353,"paw paw, il",294450.0,0.0865,0.0711,7,0.2727,0.4000,27.0,...,432643.0,0.2654,0.2023,8,0.3333,0.1429,0.2143,0.0325,-0.3857,1.0
2,202608,35765,"pisgah, al",307950.0,0.0448,0.1322,15,0.0345,0.5000,105.0,...,312969.0,-0.0256,-0.1669,18,0.1613,0.3846,0.1333,0.0644,-0.0667,1.0
3,202608,21664,"secretary, md",349900.0,0.0294,0.0218,2,0.3333,-0.5000,54.0,...,349900.0,0.0294,0.1915,2,-0.3333,-0.5000,0.0000,-0.6667,-0.2500,0.0
4,202608,37221,"nashville, tn",519900.0,-0.0295,-0.0877,235,0.0129,0.0930,64.0,...,742763.0,0.0067,0.0582,312,-0.0473,-0.0064,0.3277,-0.0861,-0.1375,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3452028,201607,8828,"helmetta, nj",218000.0,NaN,NaN,22,NaN,NaN,67.0,...,243027.0,NaN,NaN,22,NaN,NaN,0.0455,NaN,NaN,NaN
3452029,201607,30536,"ellijay, ga",249350.0,NaN,NaN,128,NaN,NaN,86.0,...,309281.0,NaN,NaN,128,NaN,NaN,0.0118,NaN,NaN,NaN
3452030,201607,37863,"pigeon forge, tn",224750.0,NaN,NaN,189,NaN,NaN,115.0,...,283587.0,NaN,NaN,226,NaN,NaN,0.2042,NaN,NaN,NaN
3452031,201607,32097,"yulee, fl",231973.0,NaN,NaN,150,NaN,NaN,69.0,...,324656.0,NaN,NaN,252,NaN,NaN,0.6923,NaN,NaN,NaN


In [237]:
august = pd.read_csv('/Users/medhasubramaniyan/Desktop/GitHub_Repos/orlando-seller-leverage-analysis/data/RDC_Inventory_Core_Metrics_Zip.csv')
august 

,month_date_yyyymm,postal_code,zip_name,median_listing_price,median_listing_price_mm,median_listing_price_yy,active_listing_count,active_listing_count_mm,active_listing_count_yy,median_days_on_market,...,average_listing_price,average_listing_price_mm,average_listing_price_yy,total_listing_count,total_listing_count_mm,total_listing_count_yy,pending_ratio,pending_ratio_mm,pending_ratio_yy,quality_flag
0,202608,78956,"schulenburg, tx",314000.0,-0.1253,0.0502,28,0.1200,0.4000,124.0,...,410433.0,-0.0986,0.2020,32,0.1228,0.3913,0.1786,0.0386,-0.0214,1
1,202608,30535,"demorest, ga",415000.0,-0.0259,0.0038,55,-0.0090,0.4865,90.0,...,461853.0,-0.0132,0.0613,68,-0.0933,0.4167,0.2545,-0.0878,-0.0428,1
2,202608,83316,"buhl, id",510000.0,0.0000,0.2146,45,-0.0110,-0.0816,79.0,...,699305.0,-0.0023,0.2881,55,-0.0833,-0.1129,0.2222,-0.1184,-0.0635,0
3,202608,94806,"san pablo, ca",577500.0,0.0226,-0.0077,71,0.2034,-0.1013,38.0,...,574916.0,0.0050,-0.0158,91,0.0581,-0.1947,0.3099,-0.1732,-0.1205,0
4,202608,37820,"new market, tn",735000.0,0.0426,0.4703,23,0.1500,0.0455,57.0,...,771742.0,-0.0668,0.5105,29,0.1154,-0.1714,0.3043,-0.0207,-0.1957,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
29046,202608,50546,"havelock, ia",265500.0,0.0000,0.0215,1,0.0000,0.0000,81.0,...,265500.0,0.0000,0.0215,2,1.0000,0.0000,1.0000,NaN,0.0000,0
29047,202608,54530,"hawkins, wi",297500.0,0.0000,2.1349,3,-0.1429,0.5000,59.0,...,244967.0,-0.0126,1.5813,4,0.1429,0.3333,0.6667,NaN,0.1667,1
29048,202608,35755,"langston, al",574500.0,0.1011,-0.2584,12,0.0000,1.0000,115.0,...,725423.0,0.0650,-0.3124,14,0.1667,1.0000,0.0833,NaN,-0.0833,1
29049,202608,69148,"lisco, ne",98700.0,0.0000,-0.7005,1,0.0000,0.0000,107.0,...,98700.0,0.0000,-0.7005,1,0.0000,0.0000,1.0000,NaN,NaN,1


## Filter the national data down to Florida, Central Florida, and Orlando (August 2026, and Historical)

In [238]:
current_data = august.copy()
past_data = historical.copy()

In [239]:
# greater central forida zips 
orlando_orange_county = [
    "32801", "32803", "32804", "32805", "32806", "32808", "32810",
    "32807", "32817", "32826", "32819", "32828", "32827", "32832", 
    "32835", "32703", "32704", "32712", "34761", "34787", "34786"
]
orlando_orange_county = list(map(int, orlando_orange_county))

seminole_county = [
    "32701", "32714", "32750", "32779", "32746", "32771", 
    "32772", "32773", "32707", "32708", "32765", "32766"
]
seminole_county = list(map(int, seminole_county))

osceola_county = [
    "34741", "34743", "34744", "34746", "34747", "34758",
    "34769", "34770", "34771", "34772"
]
osceola_county = list(map(int, osceola_county))

central_fl_tri_county = orlando_orange_county + seminole_county + osceola_county


In [240]:
def filterToFlorida(df):
    '''pass the national level data here to filter just to the state of florida '''
    florida = df.copy()
    florida[['city', 'state']] = florida['zip_name'].str.split(',', expand=True)
    florida['state'] = florida['state'].str.strip()
    florida['city'] = florida['city'].str.strip()
    florida = florida[florida['state'] == 'fl'] #we've filtered down to florida! 
    return florida 

# central_fl = orl[orl['postal_code'].isin(central_fl_tri_county)]
# orl = orl[orl['postal_code'].isin(orlando_orange_county)]
def filterToCentralFlorida(df):
        '''pass the florida level data here to filter just to central florida and further filter to orlando '''
        central_fl_df = df[df['postal_code'].isin(central_fl_tri_county)] 
        orlando_df = central_fl_df[central_fl_df['postal_code'].isin(orlando_orange_county)]
        return central_fl_df, orlando_df
   

In [241]:

fl_august = filterToFlorida(current_data)  # current florida august data 
fl_historical =  filterToFlorida(past_data) # past florida data 

central_fl_august, orl_august = filterToCentralFlorida(fl_august) #current central florida and orlando data 
central_fl_historical, orl_historical = filterToCentralFlorida(fl_historical) #past central florida and orlando data 


### Check to make sure everything has been filtered properly

In [242]:
print(f"---Current Florida State Data--- \n Unique States:{fl_august['state'].unique()}\n Date:{fl_august['month_date_yyyymm'].unique()}\n # of Cities: {fl_august['zip_name'].nunique()}\n")
print(f"---Historical Florida State Data--- \n Unique States:{fl_historical['state'].unique()}\n Date:{fl_historical['month_date_yyyymm'].unique()}\n # of Cities: {fl_historical['zip_name'].nunique()}\n")


---Current Florida State Data--- 
 Unique States:['fl']
 Date:[202608]
 # of Cities: 418

---Historical Florida State Data--- 
 Unique States:['fl']
 Date:[202608 202607 202606 202605 202604 202603 202602 202601 202512 202511
 202510 202509 202508 202507 202506 202505 202504 202503 202502 202501
 202412 202411 202410 202409 202408 202407 202406 202405 202404 202403
 202402 202401 202312 202311 202310 202309 202308 202307 202306 202305
 202304 202303 202302 202301 202212 202211 202210 202209 202208 202207
 202206 202205 202204 202203 202202 202201 202112 202111 202110 202109
 202108 202107 202106 202105 202104 202103 202102 202101 202012 202011
 202010 202009 202008 202007 202006 202005 202004 202003 202002 202001
 201912 201911 201910 201909 201908 201907 201906 201905 201904 201903
 201902 201901 201812 201811 201810 201809 201808 201807 201806 201805
 201804 201803 201802 201801 201712 201711 201710 201709 201708 201707
 201706 201705 201704 201703 201702 201701 201612 201611 201610 

In [243]:
print(f"---Current Central Florida Data--- \n Unique States:{central_fl_august['state'].unique()}\n Date:{central_fl_august['month_date_yyyymm'].unique()}\n # of Cities: {central_fl_august['zip_name'].nunique()}\n")
print(f"---Historical Central Florida Data--- \n Unique States:{central_fl_historical['state'].unique()}\n Date:{central_fl_historical['month_date_yyyymm'].unique()}\n # of Cities: {central_fl_historical['zip_name'].nunique()}\n")

---Current Central Florida Data--- 
 Unique States:['fl']
 Date:[202608]
 # of Cities: 14

---Historical Central Florida Data--- 
 Unique States:['fl']
 Date:[202608 202607 202606 202605 202604 202603 202602 202601 202512 202511
 202510 202509 202508 202507 202506 202505 202504 202503 202502 202501
 202412 202411 202410 202409 202408 202407 202406 202405 202404 202403
 202402 202401 202312 202311 202310 202309 202308 202307 202306 202305
 202304 202303 202302 202301 202212 202211 202210 202209 202208 202207
 202206 202205 202204 202203 202202 202201 202112 202111 202110 202109
 202108 202107 202106 202105 202104 202103 202102 202101 202012 202011
 202010 202009 202008 202007 202006 202005 202004 202003 202002 202001
 201912 201911 201910 201909 201908 201907 201906 201905 201904 201903
 201902 201901 201812 201811 201810 201809 201808 201807 201806 201805
 201804 201803 201802 201801 201712 201711 201710 201709 201708 201707
 201706 201705 201704 201703 201702 201701 201612 201611 2016

In [244]:
print(f"---Current Orlando Data---  \n Unique States:{orl_august['state'].unique()}\n Date:{orl_august['month_date_yyyymm'].unique()}\n # of Cities: {orl_august['zip_name'].nunique()}\n")
print(f"---Historical Orlando Data---  \n Unique States:{orl_historical['state'].unique()}\n Date:{orl_historical['month_date_yyyymm'].unique()}\n # of Cities: {orl_historical['zip_name'].nunique()}\n")

---Current Orlando Data---  
 Unique States:['fl']
 Date:[202608]
 # of Cities: 5

---Historical Orlando Data---  
 Unique States:['fl']
 Date:[202608 202607 202606 202605 202604 202603 202602 202601 202512 202511
 202510 202509 202508 202507 202506 202505 202504 202503 202502 202501
 202412 202411 202410 202409 202408 202407 202406 202405 202404 202403
 202402 202401 202312 202311 202310 202309 202308 202307 202306 202305
 202304 202303 202302 202301 202212 202211 202210 202209 202208 202207
 202206 202205 202204 202203 202202 202201 202112 202111 202110 202109
 202108 202107 202106 202105 202104 202103 202102 202101 202012 202011
 202010 202009 202008 202007 202006 202005 202004 202003 202002 202001
 201912 201911 201910 201909 201908 201907 201906 201905 201904 201903
 201902 201901 201812 201811 201810 201809 201808 201807 201806 201805
 201804 201803 201802 201801 201712 201711 201710 201709 201708 201707
 201706 201705 201704 201703 201702 201701 201612 201611 201610 201609
 2016

### Filter down to zips that Jeanine serves

In [245]:
focus_zips = [34786, 34787, 34761, 32712 , 32703, 32801] #jeanine's target market
alcime = central_fl_historical[central_fl_historical['postal_code'].isin(focus_zips)]
alcime

,month_date_yyyymm,postal_code,zip_name,median_listing_price,median_listing_price_mm,median_listing_price_yy,active_listing_count,active_listing_count_mm,active_listing_count_yy,median_days_on_market,...,average_listing_price_yy,total_listing_count,total_listing_count_mm,total_listing_count_yy,pending_ratio,pending_ratio_mm,pending_ratio_yy,quality_flag,city,state
8183,202608,32712,"apopka, fl",455000.0,-0.0246,-0.0712,259,0.0058,0.0792,72.0,...,-0.0921,411,0.0074,0.1743,0.5946,0.0101,0.1071,0.0,apopka,fl
9079,202608,34761,"ocoee, fl",438000.0,0.0368,-0.0374,137,0.0186,0.1513,68.0,...,-0.0939,215,0.0190,0.1436,0.5547,-0.0103,-0.0335,0.0,ocoee,fl
17845,202608,34787,"winter garden, fl",590000.0,0.0022,-0.0407,639,-0.0244,-0.0169,72.0,...,-0.0506,960,-0.0264,0.1622,0.5102,0.0048,0.2286,0.0,winter garden,fl
19739,202608,34786,"windermere, fl",1089000.0,-0.0200,0.1204,233,-0.0251,-0.0567,75.0,...,0.3265,310,-0.0143,-0.0520,0.3391,0.0127,0.0030,0.0,windermere,fl
19757,202608,32703,"apopka, fl",405000.0,0.0095,0.0387,169,0.0563,-0.0398,66.0,...,0.0266,263,-0.0573,-0.0993,0.5740,-0.1604,-0.0681,0.0,apopka,fl
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3424422,201607,34786,"windermere, fl",641256.0,NaN,NaN,558,NaN,NaN,81.0,...,NaN,709,NaN,NaN,0.2789,NaN,NaN,NaN,windermere,fl
3424815,201607,32703,"apopka, fl",239000.0,NaN,NaN,125,NaN,NaN,43.0,...,NaN,266,NaN,NaN,1.1325,NaN,NaN,NaN,apopka,fl
3428521,201607,34787,"winter garden, fl",360305.0,NaN,NaN,397,NaN,NaN,61.0,...,NaN,610,NaN,NaN,0.5403,NaN,NaN,NaN,winter garden,fl
3443616,201607,32801,"orlando, fl",332548.0,NaN,NaN,138,NaN,NaN,72.0,...,NaN,199,NaN,NaN,0.4457,NaN,NaN,NaN,orlando,fl


In [246]:
import duckdb



conn = duckdb.connect()
fl_zips = ', '.join(fl_historical['postal_code'].astype(str))


In [247]:
redfin = conn.sql(f""" 
                    SELECT * 
                    FROM '/Users/medhasubramaniyan/Desktop/GitHub_Repos/orlando-seller-leverage-analysis/data/redfin_housing_market_monthly_all_zips_2016_Jul_to_2026_Aug.csv' 
                    WHERE CAST("REGION NAME" AS INTEGER) IN ({fl_zips})
                     """).df()

redfin

,LAST UPDATED,FREQUENCY,PERIOD BEGIN,PERIOD END,REGION ID,REGION TYPE,REGION NAME,METRO,HOMES SOLD,HOMES SOLD YOY (%),...,MEDIAN NEW LISTING PRICE ($),MEDIAN NEW LISTING PRICE YOY (%),MEDIAN NEW LISTING PRICE PER SQ.FT. ($),MEDIAN NEW LISTING PRICE PER SQ.FT. YOY (%),MEDIAN SALE PRICE PER SQ.FT. ($),MEDIAN SALE PRICE PER SQ.FT. YOY (%),MONTHS OF SUPPLY,MONTHS OF SUPPLY YOY (%),PERCENT OFF MARKET IN TWO WEEKS (%),PERCENT OFF MARKET IN TWO WEEKS YOY (PPTS)
0,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14215,Zip,33510,"Tampa, FL metro area",77,22.22,...,359450,2.7,218.09,3.23,206.97,-3.17,3.3,-130.99,22.08,-3.95
1,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14216,Zip,33511,"Tampa, FL metro area",155,17.42,...,399000,6.4,217.64,-1.49,217.86,-0.21,3.5,-14.21,31.79,2.76
2,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14220,Zip,33523,"Tampa, FL metro area",37,-27.45,...,308595,3.31,188.87,3.85,163.88,-12.41,6.5,303.74,17.54,-8.99
3,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14222,Zip,33525,"Tampa, FL metro area",96,-13.51,...,389100,-0.23,194.37,-1.51,188.07,-3.45,4.7,66.65,12.37,-0.34
4,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14224,Zip,33527,"Tampa, FL metro area",19,-29.63,...,565000,5.62,226.23,-3.6,220.59,1.63,5.6,201.46,20.83,-1.04
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
111401,2026-09-03,Rolling 3 Months,2016-08-01,2016-10-31,13685,Zip,32628,NA,1,0,...,79900,8.71,50.76,4.14,39.39,227.12,33.7,1533.33,0,0
111402,2026-09-03,Rolling 3 Months,2016-08-01,2016-10-31,13695,Zip,32648,NA,4,33.33,...,159900,NA,140.83,NA,67.15,-64.44,9.2,408.89,0,0
111403,2026-09-03,Rolling 3 Months,2016-08-01,2016-10-31,13708,Zip,32680,NA,18,63.64,...,119300,-14.79,77.27,-19.48,65.13,-40.64,5.1,-548.28,0,0
111404,2026-09-03,Rolling 3 Months,2016-08-01,2016-10-31,14193,Zip,33471,NA,8,33.33,...,144450,11.2,77.36,-11.95,42.32,-16.92,8.4,-383.33,0,0


We have filtered down to some key dataframes

HISTORICAL DATA
- `fl_historical`
- `central_fl_historical`
- `orl_historical`

AUGUST 2026 DATA
- `fl_august`
- `central_fl_august`
- `orl_august`


`Alcime`: data for the zips Jeanine Serves

`redfin`: redfin data with more standardized metrics


# Core metrics: seller-favored, balanced, or buyer-favored

## How the two sources work together

| | Redfin | Realtor.com |
|---|---|---|
| Based on | Closed sales | Active listings |
| Timing | Lagging: reflects deals already done | Leading: changes before sales do |
| Period | Rolling 3-month windows | Single months |
| Benchmarks | Industry standards apply | No standard benchmarks |
| Role | Sets the label | Confirms the label, signals direction |

Each zip gets three outputs:
1. **Label:** Seller-favored / Balanced / Buyer-favored (Redfin benchmarks)
2. **Confidence:** High / Mixed (do both sources agree?)
3. **Outlook:** Heating up / Steady / Cooling (leading signals)

## Joining the data

- Key: zip (as string) + Redfin period end date.
- Realtor.com monthly values are averaged over the same 3 months as each
  Redfin window (e.g. Redfin "Aug 2026" = average of Realtor.com Jun, Jul,
  Aug 2026).
- Check after joining: row count, and every one of Jeanine's zips present in
  both sources for each comparison window.

## Dimension 1: Supply

**Primary (Redfin):** `MONTHS OF SUPPLY`
Check: ≈ INVENTORY ÷ (HOMES SOLD ÷ 3). Use INVENTORY, not ACTIVE LISTINGS.

- Seller-favored: under 4 months
- Balanced: 4–6 months
- Buyer-favored: over 6 months
(Benchmark source: [cite]. Redfin's reporting uses 4–5 months as balanced.)

**Supporting (Realtor.com):** `pending_ratio` (pending ÷ active listings)
Should move opposite to months of supply.

**Leading (both):**
- Supply flow: Redfin `NEW LISTINGS ÷ HOMES SOLD`. Above 1 = inventory
  building.
- Realtor.com `new_listing_count` vs. same month last year: a monthly early
  read on new supply.

## Dimension 2: Speed

**Primary (Redfin):** `MEDIAN DAYS ON MARKET` (sold homes)

- Seller-favored: under 30 days
- Balanced: 30–60 days
- Buyer-favored: over 60 days

**Supporting:**
- Redfin `PERCENT OFF MARKET IN TWO WEEKS`: higher = hotter demand.
- Realtor.com `median_days_on_market` (listings): should move in the same
  direction as Redfin's.

**Leading:** Redfin `PENDING SALES` vs. a year earlier. Pending sales lead
closed sales by 1–2 months.

## Dimension 3: Price leverage

**Primary (Redfin):** `AVERAGE SALE TO LIST RATIO`

- Seller-favored: 99% or higher
- Balanced: 95% to under 99%
- Buyer-favored: under 95%
(Balanced-band cutoff is my choice; see decision log.)

**Supporting:** Redfin `SHARE SOLD ABOVE ORIGINAL LIST`. The client-friendly
version ("X% of homes here sold over asking").

**Leading (Realtor.com):** `price_reduced_share`. Sellers cut prices before
they sell, so rising price cuts usually show up before sale-to-list falls.
This is the strongest early-warning signal across both sources.

**Note:** Redfin's sale-to-list compares the sale price to the *final* list
price, after any cuts. A zip can show 98% sale-to-list while 40% of
listings had to cut first. The two together give the full picture:
how often sellers came down (Realtor.com) and how far below the final
price they closed (Redfin).

## Dimension 4: Price trend

- Redfin `MEDIAN SALE PRICE PER SQ.FT.` vs. a year earlier: what buyers
  actually paid.
- Realtor.com `median_listing_price_per_square_foot` vs. a year earlier:
  what sellers are asking.
- Per square foot is used instead of median price, which is distorted when
  the mix of homes changes.
- If asking prices rise while sale prices are flat, sellers are
  overpricing for the market. Useful for listing appointments.

## Turning metrics into outputs

**1. Label (the map)**
Each primary metric (months of supply, days on market, sale-to-list) votes
using its benchmarks. Label = majority vote. If all three disagree,
months of supply decides.

**2. Confidence**
For each dimension, check whether the Realtor.com supporting signal agrees
with the Redfin primary metric (same direction vs. the same window last
year).
- High: all three primary metrics agree, and Realtor.com agrees on at
  least 2 of 3 dimensions.
- Mixed: anything else.

**3. Outlook**
Based on leading signals vs. the same window last year:
- Cooling: price reduced share rising, pending sales falling, supply flow
  above 1. (2 of 3 = Cooling)
- Heating up: the reverse. (2 of 3 = Heating up)
- Steady: otherwise.

Displayed together, e.g.: "Balanced (High confidence), Cooling."

**4. Change since 2022 and 2019 (the bar chart)**
- Headline: change in months of supply, Aug 2022 window → Aug 2026 window.
- Also shown: change in label (e.g. Seller-favored → Balanced).
- 2022 = peak market (distance from the frenzy).
- 2019 = normal market (has the zip normalized or overcorrected?).
- History by source: Realtor.com monthly data and the Redfin Key Metrics
  file cover 2016 onward. If the All Metrics file only starts in 2023,
  2022 months of supply and sale-to-list are reconstructed from the YoY
  columns (approximate); 2019 comparisons use days on market, pending
  sales, new listings, and price reduced share.

## Limitations

- Redfin periods are rolling 3-month windows; labels react to change more
  slowly than single months.
- Redfin sale-to-list measures against the final list price, not the
  original; price cuts before sale are captured separately.
- Low-volume zips are excluded or flagged (see decision log for cutoff).
  Redfin suppresses some values in small zips.
- Reconstructed 2022 values are approximate.
- All property types are combined; condo-heavy zips like 32801 are compared
  against single-family-heavy zips. Property-type split is v2.
- Zips don't match neighborhoods exactly.
- Benchmarks are industry rules of thumb; cutoffs and sources are in the
  decision log.

# Data Profiling

now we enter the cleaning -> eda -> further cleaning process

In [248]:
orl_historical.reset_index(inplace=True, drop=True)
#one row = one zip code within the greater orlando area's monthly data 0 median price, pending listings, tital listings, etc
# why are some like median_listing_price_yy or median_listing_price_mm tiny decimals? 

## Data Snapshot - Clean up and explore the different columns, shape, and rows

### Orlando Realtor.com Data

In [249]:
orl_historical.info() # is the august data folded in to the historical data? 
orl_historical.shape
orl_historical.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2440 entries, 0 to 2439
Data columns (total 48 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   month_date_yyyymm                        2440 non-null   int64  
 1   postal_code                              2440 non-null   int64  
 2   zip_name                                 2440 non-null   object 
 3   median_listing_price                     2440 non-null   float64
 4   median_listing_price_mm                  2200 non-null   float64
 5   median_listing_price_yy                  2200 non-null   float64
 6   active_listing_count                     2440 non-null   int64  
 7   active_listing_count_mm                  2200 non-null   float64
 8   active_listing_count_yy                  2200 non-null   float64
 9   median_days_on_market                    2440 non-null   float64
 10  median_days_on_market_mm                 2200 no

,month_date_yyyymm,postal_code,median_listing_price,median_listing_price_mm,median_listing_price_yy,active_listing_count,active_listing_count_mm,active_listing_count_yy,median_days_on_market,median_days_on_market_mm,...,average_listing_price,average_listing_price_mm,average_listing_price_yy,total_listing_count,total_listing_count_mm,total_listing_count_yy,pending_ratio,pending_ratio_mm,pending_ratio_yy,quality_flag
count,2440.000000,2440.0000,2.440000e+03,2200.000000,2200.000000,2440.000000,2200.000000,2200.000000,2440.000000,2200.000000,...,2.440000e+03,2200.000000,2200.000000,2440.000000,2200.000000,2200.000000,2440.000000,2200.000000,2200.000000,2200.000000
mean,202114.713115,33098.8500,4.090252e+05,0.004972,0.063680,119.898770,0.011006,0.131668,54.690164,0.020831,...,5.239252e+05,0.005141,0.063295,197.832787,0.002963,0.022802,0.876738,-0.004957,-0.045738,0.201364
std,295.645630,706.3388,1.883416e+05,0.051335,0.136002,97.945011,0.140342,0.592962,16.925830,0.190799,...,3.466270e+05,0.052981,0.151993,140.240451,0.086164,0.281966,0.662955,0.269599,0.688369,0.401110
min,201607.000000,32703.0000,8.532500e+04,-0.373700,-0.430800,8.000000,-0.500000,-0.813700,11.000000,-0.596800,...,1.021040e+05,-0.352900,-0.639200,23.000000,-0.281100,-0.576600,0.101300,-2.258800,-3.262800,0.000000
25%,201901.000000,32804.7500,2.887438e+05,-0.018300,-0.022075,52.000000,-0.065200,-0.200000,42.000000,-0.095350,...,3.187818e+05,-0.020625,-0.030825,107.000000,-0.049325,-0.152500,0.445150,-0.084150,-0.270025,0.000000
50%,202107.500000,32813.5000,3.722830e+05,0.000100,0.044800,99.000000,0.000000,0.016750,53.000000,0.025450,...,4.431020e+05,0.002550,0.036750,171.000000,-0.001900,-0.007750,0.666250,-0.001050,-0.039050,0.000000
75%,202402.000000,32829.0000,4.895500e+05,0.024800,0.131750,148.000000,0.067075,0.341200,66.000000,0.123300,...,5.978458e+05,0.028100,0.131650,235.000000,0.048700,0.150950,1.078950,0.077575,0.150175,0.000000
max,202608.000000,34787.0000,1.369162e+06,0.457000,0.847600,657.000000,1.734000,6.298200,130.000000,1.240000,...,2.514529e+06,0.673900,0.858400,1007.000000,0.627100,2.485300,5.532300,2.264700,4.212000,1.000000


In [250]:
orl_historical.rename(columns={'month_date_yyyymm':'month_year'}, inplace=True)


/var/folders/cm/46nnvpl978b_vb10b1dp7rp80000gn/T/ipykernel_39477/1807519895.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  orl_historical.rename(columns={'month_date_yyyymm':'month_year'}, inplace=True)


In [251]:
# lets create a function that spits out month_year as 2 columns: month and year 
def split_monthYear(month_year):
    ''' create a function that spits out month_year as 2 columns: month and year ''' 
    string_monthYear = month_year.astype(str)
    year = string_monthYear.str[:4].astype(int)
    month = string_monthYear.str[4:].astype(int)
    return pd.DataFrame({'year': year, 'month': month})

orl_historical[['year', 'month']] = split_monthYear(orl_historical['month_year'])
orl_historical

/var/folders/cm/46nnvpl978b_vb10b1dp7rp80000gn/T/ipykernel_39477/1295504457.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  orl_historical[['year', 'month']] = split_monthYear(orl_historical['month_year'])
/var/folders/cm/46nnvpl978b_vb10b1dp7rp80000gn/T/ipykernel_39477/1295504457.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  orl_historical[['year', 'month']] = split_monthYear(orl_historical['month_year'])


,month_year,postal_code,zip_name,median_listing_price,median_listing_price_mm,median_listing_price_yy,active_listing_count,active_listing_count_mm,active_listing_count_yy,median_days_on_market,...,total_listing_count_mm,total_listing_count_yy,pending_ratio,pending_ratio_mm,pending_ratio_yy,quality_flag,city,state,year,month
0,202608,32828,"orlando, fl",489450.0,-0.0056,0.0136,187,0.0418,0.2897,65.0,...,0.0000,0.3194,0.3316,-0.0723,0.0212,0.0,orlando,fl,2026,8
1,202608,32712,"apopka, fl",455000.0,-0.0246,-0.0712,259,0.0058,0.0792,72.0,...,0.0074,0.1743,0.5946,0.0101,0.1071,0.0,apopka,fl,2026,8
2,202608,34761,"ocoee, fl",438000.0,0.0368,-0.0374,137,0.0186,0.1513,68.0,...,0.0190,0.1436,0.5547,-0.0103,-0.0335,0.0,ocoee,fl,2026,8
3,202608,32803,"orlando, fl",570000.0,-0.0150,0.1692,88,-0.0112,-0.1927,54.0,...,0.0246,-0.1554,0.4091,0.0439,0.0054,0.0,orlando,fl,2026,8
4,202608,32835,"orlando, fl",249950.0,-0.0002,0.0161,218,-0.1138,-0.2829,80.0,...,-0.0586,-0.2178,0.2615,0.0724,0.1167,0.0,orlando,fl,2026,8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2435,201607,32827,"orlando, fl",509560.0,NaN,NaN,135,NaN,NaN,72.0,...,NaN,NaN,0.3407,NaN,NaN,NaN,orlando,fl,2016,7
2436,201607,32826,"orlando, fl",250975.0,NaN,NaN,42,NaN,NaN,35.0,...,NaN,NaN,1.1205,NaN,NaN,NaN,orlando,fl,2016,7
2437,201607,32806,"orlando, fl",323675.0,NaN,NaN,127,NaN,NaN,44.0,...,NaN,NaN,0.7312,NaN,NaN,NaN,orlando,fl,2016,7
2438,201607,32828,"orlando, fl",282350.0,NaN,NaN,198,NaN,NaN,42.0,...,NaN,NaN,0.9444,NaN,NaN,NaN,orlando,fl,2016,7


In [252]:
# Creata column that merges month and year into a real datetime column
orl_historical['date'] = pd.to_datetime(dict(year=orl_historical['year'], month=orl_historical['month'], day=1)).dt.to_period('M')
orl_historical['date'] 

/var/folders/cm/46nnvpl978b_vb10b1dp7rp80000gn/T/ipykernel_39477/3625488107.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  orl_historical['date'] = pd.to_datetime(dict(year=orl_historical['year'], month=orl_historical['month'], day=1)).dt.to_period('M')


0       2026-08
1       2026-08
2       2026-08
3       2026-08
4       2026-08
         ...   
2435    2016-07
2436    2016-07
2437    2016-07
2438    2016-07
2439    2016-07
Name: date, Length: 2440, dtype: period[M]

In [253]:
orl_historical['postal_code'] =  orl_historical['postal_code'].astype(str)

/var/folders/cm/46nnvpl978b_vb10b1dp7rp80000gn/T/ipykernel_39477/2273046015.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  orl_historical['postal_code'] =  orl_historical['postal_code'].astype(str)


In [254]:
orl_historical.duplicated().sum() # no duplicate rows!

np.int64(0)

In [255]:
orl_historical.isna().sum()
# no null zips, or dates 

month_year                                    0
postal_code                                   0
zip_name                                      0
median_listing_price                          0
median_listing_price_mm                     240
median_listing_price_yy                     240
active_listing_count                          0
active_listing_count_mm                     240
active_listing_count_yy                     240
median_days_on_market                         0
median_days_on_market_mm                    240
median_days_on_market_yy                    240
new_listing_count                             0
new_listing_count_mm                        240
new_listing_count_yy                        240
price_increased_count                         0
price_increased_count_mm                   1245
price_increased_count_yy                   1140
price_increased_share                         0
price_increased_share_mm                    240
price_increased_share_yy                

In [256]:
orl_historical[orl_historical['pending_ratio_mm'].isna()].head(50) #interesting, its for 06-2017 and before 
# technically, if the main metric isnt null the MoM could be calculated, YoY cant bc theres no 2016 data

,month_year,postal_code,zip_name,median_listing_price,median_listing_price_mm,median_listing_price_yy,active_listing_count,active_listing_count_mm,active_listing_count_yy,median_days_on_market,...,total_listing_count_yy,pending_ratio,pending_ratio_mm,pending_ratio_yy,quality_flag,city,state,year,month,date
2200,201706,32819,"orlando, fl",400000.0,NaN,NaN,220,NaN,NaN,73.0,...,NaN,0.4875,NaN,NaN,NaN,orlando,fl,2017,6,2017-06
2201,201706,32832,"orlando, fl",350975.0,NaN,NaN,155,NaN,NaN,62.0,...,NaN,0.5275,NaN,NaN,NaN,orlando,fl,2017,6,2017-06
2202,201706,32805,"orlando, fl",127493.0,NaN,NaN,24,NaN,NaN,60.0,...,NaN,0.9375,NaN,NaN,NaN,orlando,fl,2017,6,2017-06
2203,201706,34787,"winter garden, fl",377860.0,NaN,NaN,414,NaN,NaN,52.0,...,NaN,0.7195,NaN,NaN,NaN,winter garden,fl,2017,6,2017-06
2204,201706,32712,"apopka, fl",332815.0,NaN,NaN,168,NaN,NaN,49.0,...,NaN,0.8299,NaN,NaN,NaN,apopka,fl,2017,6,2017-06
2205,201706,32826,"orlando, fl",263500.0,NaN,NaN,35,NaN,NaN,30.0,...,NaN,1.4928,NaN,NaN,NaN,orlando,fl,2017,6,2017-06
2206,201706,32827,"orlando, fl",473250.0,NaN,NaN,152,NaN,NaN,69.0,...,NaN,0.3168,NaN,NaN,NaN,orlando,fl,2017,6,2017-06
2207,201706,32703,"apopka, fl",249900.0,NaN,NaN,114,NaN,NaN,48.0,...,NaN,1.0441,NaN,NaN,NaN,apopka,fl,2017,6,2017-06
2208,201706,34761,"ocoee, fl",293748.0,NaN,NaN,128,NaN,NaN,46.0,...,NaN,1.0824,NaN,NaN,NaN,ocoee,fl,2017,6,2017-06
2209,201706,34786,"windermere, fl",696000.0,NaN,NaN,490,NaN,NaN,79.0,...,NaN,0.3361,NaN,NaN,NaN,windermere,fl,2017,6,2017-06


In [257]:
orl_historical.dtypes #postal code should be a string 

month_year                                     int64
postal_code                                   object
zip_name                                      object
median_listing_price                         float64
median_listing_price_mm                      float64
median_listing_price_yy                      float64
active_listing_count                           int64
active_listing_count_mm                      float64
active_listing_count_yy                      float64
median_days_on_market                        float64
median_days_on_market_mm                     float64
median_days_on_market_yy                     float64
new_listing_count                              int64
new_listing_count_mm                         float64
new_listing_count_yy                         float64
price_increased_count                          int64
price_increased_count_mm                     float64
price_increased_count_yy                     float64
price_increased_share                        f

Columns needed from realtor.com data:

Dimesnions 1:
- pending_ratio (why are mm andd yy 240 nulls) 
- new_listing_count (why are mm andd yy 240 nulls)  

Dimension 2: median_days_on_market (why are mm andd yy 240 nulls)  

Dimension 3: price_reduced_share 

Dimension 4: median_listing_price_per_square_foot


In [258]:
realtor_columns = ['date', 'postal_code', 'zip_name','pending_ratio', 'new_listing_count', 'median_days_on_market', 'price_reduced_share', 'median_listing_price_per_square_foot']
orl_realtor = orl_historical[realtor_columns]
orl_realtor.isna().sum()

date                                    0
postal_code                             0
zip_name                                0
pending_ratio                           0
new_listing_count                       0
median_days_on_market                   0
price_reduced_share                     0
median_listing_price_per_square_foot    0
dtype: int64

In [259]:
orl_realtor['date'].min()
orl_realtor['date'].max()

Period('2026-08', 'M')

### Orlando Redfin Data

In [260]:
redfin

,LAST UPDATED,FREQUENCY,PERIOD BEGIN,PERIOD END,REGION ID,REGION TYPE,REGION NAME,METRO,HOMES SOLD,HOMES SOLD YOY (%),...,MEDIAN NEW LISTING PRICE ($),MEDIAN NEW LISTING PRICE YOY (%),MEDIAN NEW LISTING PRICE PER SQ.FT. ($),MEDIAN NEW LISTING PRICE PER SQ.FT. YOY (%),MEDIAN SALE PRICE PER SQ.FT. ($),MEDIAN SALE PRICE PER SQ.FT. YOY (%),MONTHS OF SUPPLY,MONTHS OF SUPPLY YOY (%),PERCENT OFF MARKET IN TWO WEEKS (%),PERCENT OFF MARKET IN TWO WEEKS YOY (PPTS)
0,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14215,Zip,33510,"Tampa, FL metro area",77,22.22,...,359450,2.7,218.09,3.23,206.97,-3.17,3.3,-130.99,22.08,-3.95
1,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14216,Zip,33511,"Tampa, FL metro area",155,17.42,...,399000,6.4,217.64,-1.49,217.86,-0.21,3.5,-14.21,31.79,2.76
2,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14220,Zip,33523,"Tampa, FL metro area",37,-27.45,...,308595,3.31,188.87,3.85,163.88,-12.41,6.5,303.74,17.54,-8.99
3,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14222,Zip,33525,"Tampa, FL metro area",96,-13.51,...,389100,-0.23,194.37,-1.51,188.07,-3.45,4.7,66.65,12.37,-0.34
4,2026-09-03,Rolling 3 Months,2025-08-01,2025-10-31,14224,Zip,33527,"Tampa, FL metro area",19,-29.63,...,565000,5.62,226.23,-3.6,220.59,1.63,5.6,201.46,20.83,-1.04
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
111401,2026-09-03,Rolling 3 Months,2016-08-01,2016-10-31,13685,Zip,32628,NA,1,0,...,79900,8.71,50.76,4.14,39.39,227.12,33.7,1533.33,0,0
111402,2026-09-03,Rolling 3 Months,2016-08-01,2016-10-31,13695,Zip,32648,NA,4,33.33,...,159900,NA,140.83,NA,67.15,-64.44,9.2,408.89,0,0
111403,2026-09-03,Rolling 3 Months,2016-08-01,2016-10-31,13708,Zip,32680,NA,18,63.64,...,119300,-14.79,77.27,-19.48,65.13,-40.64,5.1,-548.28,0,0
111404,2026-09-03,Rolling 3 Months,2016-08-01,2016-10-31,14193,Zip,33471,NA,8,33.33,...,144450,11.2,77.36,-11.95,42.32,-16.92,8.4,-383.33,0,0


In [261]:
redfin.shape
redfin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111406 entries, 0 to 111405
Data columns (total 36 columns):
 #   Column                                       Non-Null Count   Dtype         
---  ------                                       --------------   -----         
 0   LAST UPDATED                                 111406 non-null  datetime64[us]
 1   FREQUENCY                                    111406 non-null  object        
 2   PERIOD BEGIN                                 111406 non-null  datetime64[us]
 3   PERIOD END                                   111406 non-null  datetime64[us]
 4   REGION ID                                    111406 non-null  int64         
 5   REGION TYPE                                  111406 non-null  object        
 6   REGION NAME                                  111406 non-null  object        
 7   METRO                                        111406 non-null  object        
 8   HOMES SOLD                                   111406 non-null  ob

In [262]:
redfin.isna().sum().sum() # this is 0 becuase it uses NA instead of null ? but shouldnt it still be counted as null?
(redfin == 'NA').sum().sort_values()  # ohhh, its bc theyre strings
# convert NA to null 
redfin = redfin.replace('NA', pd.NA)
redfin.isna().sum().sum() # over 68k nulls!
redfin.isna().sum()


LAST UPDATED                                      0
FREQUENCY                                         0
PERIOD BEGIN                                      0
PERIOD END                                        0
REGION ID                                         0
REGION TYPE                                       0
REGION NAME                                       0
METRO                                          6730
HOMES SOLD                                     1868
HOMES SOLD YOY (%)                             3051
MEDIAN SALE PRICE NSA ($)                      1868
MEDIAN SALE PRICE NSA YOY (%)                  3051
MEDIAN DAYS ON MARKET (DAYS)                   2115
MEDIAN DAYS ON MARKET YOY (DAYS)               3445
AVERAGE SALE TO LIST RATIO (%)                 1877
AVERAGE SALE TO LIST RATIO YOY (PPTS)          3062
SHARE SOLD ABOVE ORIGINAL LIST (%)             1868
SHARE SOLD ABOVE ORIGINAL LIST YOY (PPTS)      3051
NEW LISTINGS                                   1457
NEW LISTINGS

In [263]:
redfin.dtypes # mostly object types

LAST UPDATED                                   datetime64[us]
FREQUENCY                                              object
PERIOD BEGIN                                   datetime64[us]
PERIOD END                                     datetime64[us]
REGION ID                                               int64
REGION TYPE                                            object
REGION NAME                                            object
METRO                                                  object
HOMES SOLD                                             object
HOMES SOLD YOY (%)                                     object
MEDIAN SALE PRICE NSA ($)                              object
MEDIAN SALE PRICE NSA YOY (%)                          object
MEDIAN DAYS ON MARKET (DAYS)                           object
MEDIAN DAYS ON MARKET YOY (DAYS)                       object
AVERAGE SALE TO LIST RATIO (%)                         object
AVERAGE SALE TO LIST RATIO YOY (PPTS)                  object
SHARE SO

columns needed from redfin data

Dimension 1:
- `MONTHS OF SUPPLY`
Check: ≈ INVENTORY ÷ (HOMES SOLD ÷ 3). Use INVENTORY, not ACTIVE LISTINGS.
- `NEW LISTINGS ÷ HOMES SOLD` 


Dimension 2:
- `MEDIAN DAYS ON MARKET`
- `PERCENT OFF MARKET IN TWO WEEKS`
- `PENDING SALES`

Dimension 3:
- `AVERAGE SALE TO LIST RATIO`
- `SHARE SOLD ABOVE ORIGINAL LIST`

Dimension 4: 
- `MEDIAN SALE PRICE PER SQ.FT`. 
- `MEDIAN SALE PRICE PER SQ.FT`.

In [264]:
redfin_cols = ['FREQUENCY', 'PERIOD BEGIN','PERIOD END','REGION ID', 'REGION NAME','METRO', 'MONTHS OF SUPPLY', 'NEW LISTINGS', 'HOMES SOLD', 'INVENTORY', 'MEDIAN DAYS ON MARKET (DAYS)', 'PERCENT OFF MARKET IN TWO WEEKS (%)', 'PENDING SALES', 'AVERAGE SALE TO LIST RATIO (%)', 'SHARE SOLD ABOVE ORIGINAL LIST (%)', 'MEDIAN SALE PRICE PER SQ.FT. ($)']
redfin_filtered = redfin[redfin_cols] # cut the unnecessary columns

In [265]:
redfin_filtered.dtypes

FREQUENCY                                      object
PERIOD BEGIN                           datetime64[us]
PERIOD END                             datetime64[us]
REGION ID                                       int64
REGION NAME                                    object
METRO                                          object
MONTHS OF SUPPLY                               object
NEW LISTINGS                                   object
HOMES SOLD                                     object
INVENTORY                                      object
MEDIAN DAYS ON MARKET (DAYS)                   object
PERCENT OFF MARKET IN TWO WEEKS (%)            object
PENDING SALES                                  object
AVERAGE SALE TO LIST RATIO (%)                 object
SHARE SOLD ABOVE ORIGINAL LIST (%)             object
MEDIAN SALE PRICE PER SQ.FT. ($)               object
dtype: object

In [266]:
num_cols = redfin_filtered.columns.to_list()
num_cols = num_cols[6:]
num_cols

redfin_filtered[num_cols] = redfin_filtered[num_cols].apply(pd.to_numeric)


/var/folders/cm/46nnvpl978b_vb10b1dp7rp80000gn/T/ipykernel_39477/640656121.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  redfin_filtered[num_cols] = redfin_filtered[num_cols].apply(pd.to_numeric)


In [267]:
redfin_filtered.dtypes

FREQUENCY                                      object
PERIOD BEGIN                           datetime64[us]
PERIOD END                             datetime64[us]
REGION ID                                       int64
REGION NAME                                    object
METRO                                          object
MONTHS OF SUPPLY                              float64
NEW LISTINGS                                  float64
HOMES SOLD                                    float64
INVENTORY                                     float64
MEDIAN DAYS ON MARKET (DAYS)                  float64
PERCENT OFF MARKET IN TWO WEEKS (%)           float64
PENDING SALES                                 float64
AVERAGE SALE TO LIST RATIO (%)                float64
SHARE SOLD ABOVE ORIGINAL LIST (%)            float64
MEDIAN SALE PRICE PER SQ.FT. ($)              float64
dtype: object

In [268]:
orl_zips = [str(z) for z in orlando_orange_county]
redfin_orl = redfin_filtered[redfin_filtered['REGION NAME'].isin(orl_zips)].copy()
redfin_orl.isna().sum()

FREQUENCY                              0
PERIOD BEGIN                           0
PERIOD END                             0
REGION ID                              0
REGION NAME                            0
METRO                                  0
MONTHS OF SUPPLY                       0
NEW LISTINGS                           0
HOMES SOLD                             0
INVENTORY                              0
MEDIAN DAYS ON MARKET (DAYS)           0
PERCENT OFF MARKET IN TWO WEEKS (%)    0
PENDING SALES                          0
AVERAGE SALE TO LIST RATIO (%)         0
SHARE SOLD ABOVE ORIGINAL LIST (%)     0
MEDIAN SALE PRICE PER SQ.FT. ($)       0
dtype: int64

In [269]:
redfin_orl['date'] = redfin_orl['PERIOD END'].dt.to_period('M')
redfin_orl

,FREQUENCY,PERIOD BEGIN,PERIOD END,REGION ID,REGION NAME,METRO,MONTHS OF SUPPLY,NEW LISTINGS,HOMES SOLD,INVENTORY,MEDIAN DAYS ON MARKET (DAYS),PERCENT OFF MARKET IN TWO WEEKS (%),PENDING SALES,AVERAGE SALE TO LIST RATIO (%),SHARE SOLD ABOVE ORIGINAL LIST (%),MEDIAN SALE PRICE PER SQ.FT. ($),date
210,Rolling 3 Months,2025-08-01,2025-10-31,13719,32703,"Orlando, FL metro area",3.2,250.0,232.0,240.0,51.0,20.00,255.0,98.62,8.19,212.89,2025-10
214,Rolling 3 Months,2025-08-01,2025-10-31,13726,32712,"Orlando, FL metro area",4.6,300.0,231.0,350.0,57.0,23.85,260.0,98.61,12.12,208.68,2025-10
236,Rolling 3 Months,2025-08-01,2025-10-31,13792,32801,"Orlando, FL metro area",10.7,90.0,44.0,154.0,83.0,13.46,52.0,95.02,0.00,268.99,2025-10
237,Rolling 3 Months,2025-08-01,2025-10-31,13794,32803,"Orlando, FL metro area",3.8,135.0,100.0,125.0,36.0,35.58,104.0,96.80,7.00,322.22,2025-10
238,Rolling 3 Months,2025-08-01,2025-10-31,13795,32804,"Orlando, FL metro area",3.2,110.0,112.0,118.0,52.0,31.62,117.0,96.40,5.36,333.94,2025-10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
111028,Rolling 3 Months,2016-07-01,2016-09-30,13821,32832,"Orlando, FL metro area",3.4,198.0,165.0,181.0,43.0,36.02,186.0,97.62,12.12,128.19,2016-09
111030,Rolling 3 Months,2016-07-01,2016-09-30,13824,32835,"Orlando, FL metro area",3.1,252.0,244.0,249.0,46.0,29.39,245.0,97.39,12.30,105.81,2016-09
111055,Rolling 3 Months,2016-07-01,2016-09-30,14733,34761,"Orlando, FL metro area",2.3,194.0,223.0,166.0,45.0,33.92,227.0,97.96,17.49,111.83,2016-09
111060,Rolling 3 Months,2016-07-01,2016-09-30,14743,34786,"Orlando, FL metro area",6.7,335.0,252.0,552.0,86.0,15.75,292.0,96.70,5.95,142.43,2016-09


In [270]:
print(redfin_orl['PERIOD BEGIN'].min()) #7-2016
redfin_orl.isna().sum()
print(redfin_orl['PERIOD BEGIN'].max()) #06-2026
redfin_orl['REGION NAME'].nunique() #20 
orl_realtor['postal_code'].nunique() # also 20



2016-07-01 00:00:00
2026-06-01 00:00:00


20

In [271]:
focus_zips = [str(z) for z in focus_zips]
alcime_redfin = redfin_orl[redfin_orl['REGION NAME'].isin(focus_zips)]
alcime_redfin['REGION NAME'].nunique()
alcime_redfin['PERIOD BEGIN'].min()
#all of jeanine's zips survived starting at 1-1-2022

Timestamp('2016-07-01 00:00:00')

In [272]:
#lets export cleaned data 
redfin_orl.to_csv('orlando_redfin.csv', index=False)
orl_realtor.to_csv('orlando_realtor.csv', index=False)


## EDA - (distributions, seasonality, join realtor to redfin and finalize data)

- redfin_orl = orlando data from redfin.com 2016-2026 
- orl_realtor = orlando data from realtor.com 2016-2026 

In [291]:
# first lets make sure they have all the same zips
redfin_zips = list(redfin_orl['REGION NAME'].unique())
realtor_zips = list(orl_realtor['postal_code'].unique())
print(set(redfin_zips) == set (realtor_zips)) #yay!

True
